src/hash.c

```diff
@@ -18,6 +18,60 @@
 #include "ccache.h"
 
 #define HASH_DELIMITER "\000cCaChE"
+#define HASH_DEBUG_DELIMITER "### "
+
+// binary input, for hashing
+char *debug_hash_bin;
+
+// text input, for debugging
+char *debug_hash_txt;
+
+void hash_debug(const char *bin, const char *txt)
+{
+	static char *hash_types = "cdp"; // common, direct, cpp
+	if (bin) {
+		debug_hash_bin = x_strdup(bin);
+		assert(debug_hash_bin[strlen(debug_hash_bin)-1] == 'X');
+		for (char *p = hash_types; *p != '\0'; p++) {
+			debug_hash_bin[strlen(debug_hash_bin)-1] = *p;
+			x_try_unlink(debug_hash_bin);
+		}
+	}
+	if (txt) {
+		debug_hash_txt = x_strdup(txt);
+		x_try_unlink(debug_hash_txt);
+	}
+}
+
+static void
+hash_binary_buffer(struct mdfour *md, const void *s, size_t len)
+{
+	mdfour_update(md, (unsigned char *)s, len);
+	if (!md->identifier || len == 0) {
+		return;
+	}
+	if (debug_hash_bin) {
+		// log to different files, for the different hash types
+		debug_hash_bin[strlen(debug_hash_bin)-1] = md->identifier;
+
+		FILE *f = fopen(debug_hash_bin, "a");
+		fwrite(s, 1, len, f);
+		fclose(f);
+	}
+}
+
+static void
+hash_debug_buffer(struct mdfour *md, const void *s, size_t len)
```

Comment:
Let's see, `hash_binary_buffer` hashes a (binary?) buffer, but `hash_debug_buffer` doesn't hash a buffer... Rename to something like `hash_note_input` instead? And I think I would prefer renaming `hash_binary_buffer` back to `hash_buffer`.

Reply:
Actually `hash_buffer` _is_ the same function as before, the other two are just internal (static) helpers. One doing the hashing and logging to the binary files, and one logging to the text file.